doc/ractor.md

```diff
@@ -0,0 +1,882 @@
+    [o, o.object_id == o2.object_id]
+  }
+  #=> [[1, true], [:sym, true], [:xyzzy, true], [\"frozen\", true], [(3/1), true], [(3+4i), true], [/regexp/, true], [C, true]]
+
+  ur = unshareable_objects = ['mutable str'.dup, [:array], {hash: true}].map{|o|
+    r << o
+    o2 = r.take
+    [o, o.object_id == o2.object_id]
+  }
+  #+> "[[\"mutable str\", false], [[:array], false], [{:hash=>true}, false]]]"
+```
+
+## Language changes to isolate unshareable objects between Ractors
+
+To isolate unshareable objects between Ractors, we introduced additional language semantics on multi-Ractor.
+
+Note that without using Ractors, these additional semantics is not needed (100% compatible with Ruby 2).
+
+### Global variables
+
+Only main Ractor (a Ractor created at starting of interpreter) can access global variables.
+
+```ruby
+  $gv = 1
+  r = Ractor.new do
+    $gv
+  end
+
+  begin
+    r.take
+  rescue Ractor::RemoteError => e
+    e.cause.message #=> 'can not access global variables from non-main Ractors'
+  end
+```
+
+### Instance variables of shareable objects
+
+Only main Ractor can access instance variables of shareable objects.
+
+```ruby
+  class C
+    @iv = 'str'
+  end
+
+  r = Ractor.new do
+    class C
+      p @iv
+    end
+  end
+
+
+  begin
+    r.take
+  rescue => e
+    e.class #=> RuntimeError
+  end
+```
+
+```ruby
+  shared = Ractor.new{}
+  shared.instance_variable_set(:@iv, 'str')
+
+  r = Ractor.new shared do |shared|
+    p shared.instance_variable_get(:@iv)
+  end
+
+  begin
+    r.take
+  rescue Ractor::RemoteError => e
+    e.cause.message
+  end
+```
+
+Note that instance variables for class/module objects are also prohibited on Ractors.
+
+### Class variables
+
+Only main Ractor can access class variables.
+
+```ruby
+  class C
+    @@cv = 'str'
+  end
+
+  r = Ractor.new do
+    class C
+      p @@cv
+    end
+  end
+
+
+  begin
+    r.take
+  rescue => e
+    e.class #=> RuntimeError
+  end
+```
+
+### Constants
+
+Only main Ractor can read constants which refer to the unshareable object.
+
+```ruby
+  class C
+    CONST = 'str'
+  end
+  r = Ractor.new do
+    C::CONST
+  end
+  begin
+    r.take
+  rescue => e
+    e.class #=> NameError
+  end
+```
+
+Only main Ractor can define constants which refer to the unshareable object.
+
+```ruby
+  class C
+  end
+  r = Ractor.new do
+    C::CONST = 'str'
+  end
+  begin
+    r.take
+  rescue => e
+    e.class
+  end
+```
+
+## Implementation note
+
+* Each Ractor has own thread, it means each Ractor has at least 1 native thread.
+* Each Ractor has own ID (`rb_ractor_t::id`).
+  * On debug mode, all unshareable objects are labeled with current Ractor's id, and it is checked to detect unshareable object leak (access an object from different Ractor) in VM.
+
+## Examples
+
+### Traditional Ring example in Actor-model
+
+```ruby
+RN = 10000
+CR = Ractor.current
+
+last_r = r = Ractor.new do
+  p Ractor.recv
+  CR << :fin
+end
+
+RN.times{
+  r = Ractor.new r do |next_r|
```

Comment:
Also this assignement to `r` bugs me at first look. I don't think this is needed?

```suggestion
  Ractor.new r do |next_r|
```